### PR TITLE
Fix leak in view tree observer for view annotations

### PR DESCRIPTION
**Pull request by kiryldz**
### Summary of changes

When removing single view annotation or actually calling complete destroy wrong ref was used for view tree observer which caused leak in some scenarios (e.g. when using fragments).
Now this is fixed by revisiting an approach to use correct moments of time to register / unregister global layout listener under the hood.

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Fix leak for view annotations caused by global layout listener.</changelog>`.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: #1020 

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 1]
```diff
@@ -68,8 +68,6 @@ internal class ViewAnnotationManagerImpl(
 
   override fun removeViewAnnotation(view: View): Boolean {
     val id = idLookupMap.remove(view) ?: return false
-    val annotation = annotationMap.remove(id) ?: return false
```
Comment (Chaoba) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:71]: So we don't need to remove the annotation from annotationMap?

Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:71]: good catch - we actually do 😄 
thanks, will fix!

**Review by tobrun — APPROVED** [review]

File: sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt [hunk 2]
```diff
@@ -36,12 +34,14 @@ class ViewAnnotationManagerTest {
...
     every { view.viewTreeObserver } returns mockk(relaxed = true)
     every { mapView.context } returns mockk()
+    every { view.addOnAttachStateChangeListener(any()) } just Runs
     every { mapboxMap.addViewAnnotation(any(), any()) } returns ExpectedFactory.createNone()
   }
 
   @After
   fun tearDown() {
     every { mapboxMap.removeViewAnnotation(any()) } returns ExpectedFactory.createNone()
+    every { view.removeOnAttachStateChangeListener(any()) } just Runs
```
Comment (axti) [inline sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt:44]: Shouldn't be included verification of this call:
`inflatedView.viewTreeObserver.removeOnGlobalLayoutListener` ?

**Comment by ank27** [comment]
tests are failing, else look good 👍

**Review by ank27 — APPROVED** [review]

File: sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt (hunk 2 above)
Comment (kiryldz) [reply sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt:44]: as current test setup is built - I guess not 😄

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 3]
```diff
@@ -212,8 +213,16 @@ internal class ViewAnnotationManagerImpl(
         }
       }
     }
-    viewAnnotation.globalLayoutListener = globalLayoutListener
-    inflatedView.viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
+    viewAnnotation.attachStateListener = object : View.OnAttachStateChangeListener {
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:216]: I'd add a comment to describe why it's here - not trivial and widely known relation between onViewAttach / onGlobalLayoutListener

**Review by yunikkk — APPROVED** [review]

**Review by Chaoba — APPROVED** [review]

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 3 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:216]: docs are added inside `ViewAnnotation` class for that field

**Pull request merged**